x/ibc/core/03-connection/keeper/handshake.go

```diff
@@ -74,54 +70,65 @@ func (k Keeper) ConnOpenTry(
 	proofConsensus []byte, // proof that chainA stored chainB's consensus state at consensus height
 	proofHeight exported.Height, // height at which relayer constructs proof of A storing connectionEnd in state
 	consensusHeight exported.Height, // latest height of chain B which chain A has stored in its chain B client
-) error {
+) (string, error) {
+	var (
+		connectionID       string
+		previousConnection types.ConnectionEnd
+		found              bool
+	)
+
+	// empty connection identifier indicates continuing a previous connection handshake
+	if previousConnectionID != "" {
+		// ensure that the previous connection exists
+		previousConnection, found = k.GetConnection(ctx, previousConnectionID)
+		if !found {
+			return "", sdkerrors.Wrapf(types.ErrInvalidConnection, "previous connection does not exist for supplied previous connectionID %s", previousConnectionID)
+		}
+
+		// ensure that the existing connection's
+		// counterparty is chainA and connection is on INIT stage.
+		// Check that existing connection versions for initialized connection is equal to compatible
+		// versions for this chain.
+		if !(previousConnection.State == types.INIT &&
+			previousConnection.Counterparty.ConnectionId == "" &&
+			bytes.Equal(previousConnection.Counterparty.Prefix.Bytes(), counterparty.Prefix.Bytes()) &&
+			previousConnection.ClientId == clientID &&
+			previousConnection.Counterparty.ClientId == counterparty.ClientId) {
```

Comment:
nit: I would actually like this to be split up a bit. Namely if lines `93-96` fail. then we cannot relay connection attempt as the previous connection is connecting to something else entirely and we should return an error such that relayer creates a completely new connection.

However, if lines `93-96` are fine but the previousConnection is now say in `TRYOPEN`, this implies the connection is connecting to the right counterparty but it has already progressed past where the relayer is. Thus, in this case we should return an error with the current STATE so that relayer can correctly advance the handshake or start relaying packets.

Reply:
This will allow relayer to take correct action based on error. I would even suggest using separate error codes to distinguish the two cases